src/scores/probability/fss_impl.py

```diff
@@ -0,0 +1,528 @@
+"""
+This module contains methods related to the FSS score
+
+For an explanation of the FSS, and implementation considerations,
+see: [Fast calculation of the Fractions Skill Score][fss_ref]
+
+[fss_ref]:
+https://www.researchgate.net/publication/269222763_Fast_calculation_of_the_Fractions_Skill_Score
+"""
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import TYPE_CHECKING, Iterable, Optional, Tuple, Union
+
+import numpy as np
+import numpy.typing as npt
+import xarray as xr
+
+from scores import utils
+from scores.typing import XarrayLike
+
+# Note: soft keyword `type` only support from >=3.10
+# "tuple" of 64 bit floats
+f8x3 = np.dtype("f8, f8, f8")
+# Note: `TypeAlias` on variables for python <=3.9
+if TYPE_CHECKING:  # pragma: no cover
+    FssDecomposed = Union[np.ArrayLike, np.DtypeLike]
+else:  # pragma: no cover
+    FssDecomposed = npt.NDArray[f8x3]
+
+
+class DimensionError(ValueError):
+    """
+    Wrapper for specific coordinate/dimension mismatch errors
+    """
+
+    # better to be explicit because python doesn't have braces for scoping its
+    # control flow
+    pass  # pylint: disable=unnecessary-pass
+
+
+class FssComputeMethod(Enum):
+    """
+    Choice of compute backend for FSS.
+    - Standard
+        - NUMPY (default)
+    - Optimized
+        - NUMBA_NATIVE (unavailable)
+        - NUMBA_PARALLEL (unavailable)
+    - Experimental
+        - RUST_NATIVE (unavailable)
+        - RUST_OCL (unavailable)
+    """
+
+    NUMPY = 1  # (default)
+    NUMBA_NATIVE = 2
+    NUMBA_PARALLEL = 3
+    RUST_NATIVE = 4
+    RUST_OCL = 5
+
+
+def fss_2d(
+    fcst: xr.DataArray,
+    obs: xr.DataArray,
+    *,  # Force keywords arguments to be keyword-only
+    threshold: np.float64,
+    window: Tuple[int, int],
+    spatial_dims: Tuple[str, str],
+    zero_padding: bool = False,
+    reduce_dims: Optional[Iterable[str]] = None,
+    preserve_dims: Optional[Iterable[str]] = None,
+    compute_method: FssComputeMethod = FssComputeMethod.NUMPY,
+    dask: str = "forbidden",  # see: `xarray.apply_ufunc` for options
+) -> XarrayLike:  # pylint: disable=too-many-locals
+    """
+    Aggregates the Fractions Skill Score (FSS) over 2-D spatial coordinates specified by
```

Comment:
Should this be "calculates"? Aggregates makes it sound like it's not actually calculating it and I end up a bit confused about whether this is the thing I want to call to get the fractions skill score.

Reply:
I guess this is what I'm intending to say: "Uses `fss_2d_single_field` to compute the fraction skills score for each 2D spatial field in the DataArray and then aggregates them over the output of `gather_dimensions`"

Reply:
Yes, that's good - better than my suggestion and better than what you had :)